File: training/analysis/drawLossEvolution.cpp
//
//  drawLossEvolution.cpp
//  xTrainingAnalysis
//
//  Created by Jeremi Niedziela on 02/11/2020.
//  Copyright © 2020 Jeremi Niedziela. All rights reserved.
//

#include "Helpers.hpp"

#include "Result.hpp"
#include "ResultsProcessor.hpp"

// Bottleneck:
//string aucsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8/aucs/";
//string resultsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8/trainingRuns/";
//string filePattern = "hlf_eflow_3_bottle_4_v";
//string filePattern = "hlf_eflow_3_bottle_5_v";
//string filePattern = "hlf_eflow_3_bottle_6_v";
//string filePattern = "hlf_eflow_3_bottle_7_v";
//string filePattern = "hlf_eflow_3_bottle_8_v";
//string filePattern = "hlf_eflow_3_bottle_9_v";
//string filePattern = "hlf_eflow_3_bottle_10_v";

// Batch size:
//string aucsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8_batchSizes/aucs/";
//string resultsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8_batchSizes/trainingRuns/";
//string filePattern = "hlf_eflow_3_bottle_7_bs_1_v";
//string filePattern = "hlf_eflow_3_bottle_7_bs_8_v";
//string filePattern = "hlf_eflow_3_bottle_7_bs_64_v";
//string filePattern = "hlf_eflow_3_bottle_7_bs_256_v";
//string filePattern = "hlf_eflow_3_bottle_7_bs_512_v";
//string filePattern = "hlf_eflow_3_bottle_7_bs_999999_v";
//string filePattern = "hlf_eflow_3_bottle_7_bs_999999_epochs_2000_v";

// Optimizers:
//string aucsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8_optimizers/aucs/";
//string resultsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8_optimizers/trainingRuns/";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_SGD_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_RMSprop_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_Adam_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_Adadelta_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_Adagrad_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_Adamax_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_Nadam_v";
//string filePattern = "hlf_eflow_3_bottle_6_optimizer_Ftrl_v";

// Losses:
string aucsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8_losses/aucs/";
string resultsPath =  "../trainingResults_noLeptonVeto_fatJets_dr0p8_losses/trainingRuns/";
//string filePattern = "hlf_eflow_4_bottle_6_loss_mean_absolute_error_v";
//string filePattern = "hlf_eflow_4_bottle_6_loss_mean_absolute_percentage_error_v";
//string filePattern = "hlf_eflow_4_bottle_6_loss_mean_squared_error_v";
//string filePattern = "hlf_eflow_4_bottle_6_loss_mean_squared_logarithmic_error_v";
//string filePattern = "hlf_eflow_4_bottle_6_loss_huber_loss_v";
//string filePattern = "hlf_eflow_4_bottle_6_loss_log_cosh_v";
string filePattern = "hlf_eflow_4_bottle_6_loss_cosine_similarity_v";

void drawGraph(const ModelStats &stats, string title)
{
  vector<Epoch> epochsInModel = stats.epochs;
  
  TGraph *trainingLoss = new TGraph();
  TGraph *validationLoss = new TGraph();
  
  int iPoint=0;
  
  for(Epoch epoch : epochsInModel){
    trainingLoss->SetPoint(iPoint, iPoint, epoch.trainingLoss);
    validationLoss->SetPoint(iPoint, iPoint, epoch.validationLoss);
    iPoint++;
  }
  
  trainingLoss->SetMarkerStyle(20);
  trainingLoss->SetMarkerSize(0.3);
  trainingLoss->SetMarkerColor(kRed+1);
  
  validationLoss->SetMarkerStyle(20);
  validationLoss->SetMarkerSize(0.3);
  validationLoss->SetMarkerColor(kGreen+1);
    
  
  trainingLoss->Draw("AP");
  validationLoss->Draw("Psame");
  
  trainingLoss->SetTitle(title.c_str());
  trainingLoss->GetXaxis()->SetTitle("Epoch");
  trainingLoss->GetYaxis()->SetTitle("Loss");
  
  trainingLoss->GetXaxis()->SetLimits(0, 400);
//  trainingLoss->GetYaxis()->SetRangeUser(0, 0.15);
  
  TLegend *legend = new TLegend(0.5, 0.7, 0.9, 0.9);
  legend->AddEntry(trainingLoss, "training loss", "p");
  legend->AddEntry(validationLoss, "validation loss", "p");
  legend->Draw();
}

int main()
{
  cout<<"Starting drawLossEvolution"<<endl;
  gStyle->SetOptStat(0);
  useCommaAsDecimalSeparator();
 
  cout<<"Creating application"<<endl;
  TApplication app("", 0, {});
  
  cout<<"Reading results from files"<<endl;
  auto resultsProcessor = make_unique<ResultsProcessor>();
  vector<ModelStats> stats = resultsProcessor->getModelStatsFromPathMarchingPatter(aucsPath, resultsPath, filePattern);
  
  resultsProcessor->sortModelsByAucAverageOverAllSignals(stats);
  
  cout<<"The best model based on average AUC over all signals: "<<stats.front().aucsFileName<<endl;
  
  cout<<"Plotting results"<<endl;
  TCanvas *canvas = new TCanvas("c1", "c1", 600, 1000);
  canvas->Divide(1, 3);
  
  canvas->cd(1);
  drawGraph(stats.front(), "The best model");
  
  canvas->cd(2);
  drawGraph(stats[stats.size()/2.], "Average model");
  
  canvas->cd(3);
  drawGraph(stats.back(), "The worst model");
  
  canvas->Update();
  
  cout<<"Running the application"<<endl;
  app.Run();
  return 0;
}

File: training/analysis/drawLossEvolution.py

from ROOT import TGraph, kRed, kGreen, TLegend, TCanvas, gStyle
import csv, glob

aucsPath = "../trainingResults_archs/aucs/"
resultsPath = "../trainingResults_archs/trainingRuns/"
filePattern = "hlf_efp_3_bottle_9_arch_60__60_loss_mean_absolute_error_batch_size_256_v"


input_path = glob.glob("../trainingResults_withConstituents/trainingRuns/efp_3_bottle_5_arch_5__5_loss_mean_absolute_error_optimizer_Adam_batchSize_256_StandardScaler_activation_elu_tiedWeights_False_epochs_200_23constituents_v*.csv")

saved_plots = []


def draw_graphs(stats):

    canvas = TCanvas("c1", "c1", 2880, 1800)
    canvas.Divide(5, 2)

    legend = TLegend(0.5, 0.7, 0.9, 0.9)

    training_loss = []
    validation_loss = []

    for i_pad, stats_for_file in enumerate(stats):

        canvas.cd(i_pad+1)

        training_loss.append(TGraph())
        validation_loss.append(TGraph())

        for i, stat in enumerate(stats_for_file):
            training_loss[-1].SetPoint(i, i, stat[2])
            validation_loss[-1].SetPoint(i, i, stat[5])

        training_loss[-1].SetMarkerStyle(20)
        training_loss[-1].SetMarkerSize(1.0)
        training_loss[-1].SetMarkerColor(kRed + 1)
        training_loss[-1].SetTitle("v"+str(i_pad))

        validation_loss[-1].SetMarkerStyle(20)
        validation_loss[-1].SetMarkerSize(1.0)
        validation_loss[-1].SetMarkerColor(kGreen + 1)

        training_loss[-1].Draw("AP")
        validation_loss[-1].Draw("Psame")

        training_loss[-1].GetXaxis().SetTitle("Epoch")
        training_loss[-1].GetYaxis().SetTitle("Loss")

        training_loss[-1].GetXaxis().SetLimits(0, 200)

        if i_pad == 0:
            legend.AddEntry(training_loss[-1], "training loss", "p")
            legend.AddEntry(validation_loss[-1], "validation loss", "p")

    legend.Draw()
    canvas.Update()

    filename = input_path[0]
    filename = filename.replace("trainingRuns", "plots")
    filename = filename.replace(".csv", ".pdf")
    canvas.SaveAs(filename)
    saved_plots.append(filename)


def get_stats_from_file(inFilePath):

    stats = []

    with open(inFilePath, newline='') as csvfile:
        spamreader = csv.reader(csvfile, delimiter=' ', quotechar='|')
        first_row = True

        for row in spamreader:
            if first_row:
                first_row = False
                continue

            values = row[0].split(",")

            epoch = float(values[0])
            accuracy = float(values[1])
            loss = float(values[2])
            learning_rate = float(values[3])
            validation_accuracy = float(values[4])
            validation_loss = float(values[5])

            stats.append((epoch, accuracy, loss, learning_rate, validation_accuracy, validation_loss))

    return stats


def get_stats():

    stats = []

    for filename in input_path:
        stats_for_file = get_stats_from_file(filename)
        stats.append(stats_for_file)

    return stats


def main():
    gStyle.SetOptStat(0)

    stats = get_stats()
    draw_graphs(stats)

    print("Saved plots:", saved_plots)


if __name__ == "__main__":
    main()




